### PR TITLE
Add and use CLPMutableForwardIndexV2 by default to improve ingestion performance and efficiency

### PR DESCRIPTION
**Pull request by jackluo923**
# Summary
`CLPMutableForwardIndexV2` is a comprehensive improvement over `CLPMutableForwardIndexV1`, offering enhancements in both performance and functionality. `CLPMutableForwardIndexV2` is also compatible with `CLPMutableForwardIndex` and `CLPMutableForwardIndexV2` is now the default CLP mutable forward index to be used within Pinot.

This PR is the first in a series of three PRs:
1. **CLPMutableForwardIndexV2**: Introduces the updated mutable forward index.
2. **CLPForwardIndexV2**: Implements an improved version of immutable forward index offering much higher compression ratio.
3. **Codec + Config Support**: Adds the necessary codec and config support to enable the toggle between V1 & V2 immutable forward indexes.

# Compatibility
`CLPMutableForwardIndexV2` is compatible with `CLPMutableForwardIndex`. A unit test validating this compatibility can be found in `CLPMutableForwardIndexTest#testStringV2()`. In the future, we may consider replacing `CLPMutableForwardIndex` with `CLPMutableForwardIndexV2`. However, keeping both implementations for now provides a safety net, allowing us to revert to the older version should any unforeseen issues arise with V2.

# Details
## Performance Improvement
The serialization/deserialization overhead is significantly lower compared to `CLPMutableForwardIndexV1`. With the upgrade to `clp-ffi-java` version `0.4.7`, we can now directly access the raw byte[] content of CLP-encoded messages, rather than operating over boxed String objects for logtype and dictVar and Long objects for encodedVar. On the forward index storage side, the composite mutable forward index now exclusively stores primitive types. For decoding, we also work with raw byte and long arrays, eliminating the need to first encode the bytes into a String/Long before passing them to CLP's decoder to unbox and decode the log message.

## Functionality Improvement
The new mutable forward index implements a composite index for string-typed columns with dynamic encoding options: 
1. Pure CLP encoding when the dictionary cardinality is below a configurable threshold.
2. CLP encoding combined with a raw string forward index when the dictionary cardinality exceeds the threshold.

Initially, CLP encoding transforms an extremely high-cardinality log message string into three data columns:
- Logtype (very low cardinality) - essentially an inferred format string of the log message
- Dictionary variables (medium cardinality) - variables with both alphabets and numbers
- Encoded variables (high cardinality) - pure fixed point and floating point numbers

The logtype and dictionary variables are dictionary-encoded, while the encoded variables are stored as longs. Notably, both `encodedVarIds` and `encodedVars` are multi-valued, but they are stored using a flattened single-value mutable forward index, along with a separate forward index to capture the end offsets for each multi-valued document. This approach is necessary because the maximum number of values per document is unknown during ingestion, unlike the existing multi-value forward index, which requires this information upfront. During the conversion from mutable to immutable forward index, the two single-value mutable indices are merged into a single immutable multi-valued forward index, as the max length is now known post ingestion (conversion time).

During ingestion, if the cardinality of either the `logtypeId` or `dictVarID` exceeds a predefined threshold, the ingestion mode switches to a raw bytes forward index for subsequent documents. Maintaining very large dictionaries is inefficient in Pinot due to memory and I/O constraints (memory-mapped). Switching to a raw bytes forward index helps avoid these issues. During reads, if the requested `docId` is in the raw forward index, the raw bytes are returned. Otherwise, the log type, dictionary variables, and encoded variables are decoded using the CLP decoder to return the original log message's bytes.

# Notes:
Writes are strictly sequential, while reads can be performed randomly. The supported append operations are:
- `setString(int docId, String value)` - Encodes the log message using CLP and invokes `appendEncodedMessage(@NotNull EncodedMessage clpEncodedMessage)`
- `appendEncodedMessage(@NotNull EncodedMessage clpEncodedMessage)`

# Future Work: 
- enable conversion between mutable -> immutable CLP forward index without re-encoding
- enable columnar conversion between mutable -> immutable CLP forward index
- utilize dynamic CLP encoding feature provided in CLPMutableForwardIndexV2

### CODE REVIEW

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/forward/CLPMutableForwardIndexV2.java [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+  public final String _columnName;
+
+  protected final EncodedMessage _clpEncodedMessage;
+  protected final MessageEncoder _clpMessageEncoder;
+  protected final MessageDecoder _clpMessageDecoder;
+
+  protected int _nextDocId = 0;
+  protected int _nextDictVarDocId = 0;
+  protected int _nextEncodedVarId = 0;
+  protected int _bytesRawFwdIndexDocIdStartOffset = Integer.MAX_VALUE;
```
Comment (deemoliu) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/forward/CLPMutableForwardIndexV2.java:114]: is this only for CLPV2? can we add a comment on how this value is dynamic updated?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/forward/CLPMutableForwardIndexV2.java [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+  public FixedByteSVMutableForwardIndex getEncodedVarOffset() {
+    return _encodedVarOffset;
+  }
+
+  public FixedByteSVMutableForwardIndex getEncodedVar() {
+    return _encodedVar;
+  }
+
+  /**
+   * Returns whether the mutable forward index is currently using dictionary encoding.
```
Comment (deemoliu) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/forward/CLPMutableForwardIndexV2.java:384]: > whether the mutable forward index is currently using dictionary encoding
whether the mutable forward index is currently using CLP forward indexes with dictionary encoding or rawFwdIndex.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/forward/CLPMutableForwardIndexV2.java (hunk 1 above)
Comment (jackluo923) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/forward/CLPMutableForwardIndexV2.java:114]: Yes, this is only used for CLPV2 (but might be used for CLPV3 in the future). Internally, we have classes which extends the functionality of this class, so some class member variables are marked as protected.

Comment (jackluo923) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/forward/CLPMutableForwardIndexV2.java:114]: The value is updated at the end of `appendEncodedMessage` method call. I will add a comment to make this clear.

**Review by deemoliu — APPROVED** [review]
LGTM

**Review by deemoliu — APPROVED** [review]

**Review by Jackie-Jiang — COMMENTED** [review]
For mutable index, v1 and v2 cannot co-exist. If we are not planning to make it configurable, let's just remove v1 and only keep one mutable index format

**Comment by jackluo923** [comment]
> For mutable index, v1 and v2 cannot co-exist. If we are not planning to make it configurable, let's just remove v1 and only keep one mutable index format

Sure, I will make the change in another PR.

**Pull request merged**